File: kwho-java/src/main/java/kwho.java
// vim:shiftwidth=4:tabstop=4:expandtab

import sun.security.krb5.PrincipalName;
import sun.security.krb5.internal.ccache.FileCredentialsCache;

public class kwho {
    public static void main(String[] args) {


        String cache = System.getenv("KRB5CCNAME");
        if (cache == null) {
            return;
        }

        // The FileCredentialsCache does not want to see the "FILE:" prefix
        cache = cache.replaceAll("^FILE:", cache);

        //assumes credendials cache of type "FILE:"
        FileCredentialsCache fcc = FileCredentialsCache.acquireInstance(null, cache);
        if (fcc == null) {
            return;
        }
        PrincipalName princ = fcc.getPrimaryPrincipal();
        String[] nameStrings = princ.getNameStrings();
        StringBuffer temp = new StringBuffer(nameStrings[0]);
        for (int i=1; i<nameStrings.length; i++) {
            temp.append("/");
            temp.append(nameStrings[i]);
        }
        System.out.println(temp);
    }
}

File: kwho-kotlin/src/main/kotlin/demo/kwho.kt
package demo

import sun.security.krb5.PrincipalName;
import sun.security.krb5.internal.ccache.FileCredentialsCache;
import kotlin.collections.joinToString
import kotlin.text.Regex
import kotlin.text.replace

fun main(args: Array<String>) {
    var cache: String? = System.getenv("KRB5CCNAME")
    // The FileCredentialsCache does not want to see the "FILE:" prefix
    cache = cache?.replace(Regex("^FILE:"), "")
    val fcc: FileCredentialsCache? = FileCredentialsCache.acquireInstance(null, cache)
    if (fcc == null) {
        return
    }
    val princ: PrincipalName = fcc.primaryPrincipal
    val nameStrings: Array<out String> = princ.nameStrings
    val output: String = nameStrings.joinToString("/")
    println(output)
}
